src/decay.c

```diff
@@ -175,3 +175,123 @@ decay_maybe_advance_epoch(decay_t *decay, nstime_t *new_time,
 
 	return true;
 }
+
+static inline size_t
+decay_npurge_after_interval(decay_t *decay, size_t interval) {
+	size_t i;
+	uint64_t sum = 0;
+	for (i = 0; i < interval; i++) {
+		sum += decay->backlog[i] * h_steps[i];
+	}
+	for (; i < SMOOTHSTEP_NSTEPS; i++) {
+		sum += decay->backlog[i] *
+		    (h_steps[i] - h_steps[i - interval]);
+	}
+
+	return (size_t)(sum >> SMOOTHSTEP_BFP);
+}
+
+static uint64_t
+arena_decay_compute_purge_interval_impl(tsdn_t *tsdn, decay_t *decay,
+    ecache_t *ecache, const purge_interval_opts_t *opts) {
+	if (malloc_mutex_trylock(tsdn, &decay->mtx)) {
```

Comment:
Belongs in the background thread code; the `decay_t` morally doesn't know about its mutex, and this is a backgronud thread locking optimization to begin with.